components/external_storage/export/src/export.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! To use External storage with protobufs as an application, import this module.
+//! external_storage contains the actual library code
+//! Cloud provider backends are under components/cloud
+use std::io;
+use std::path::Path;
+use std::time::Instant;
+
+#[cfg(feature = "cloud-aws")]
+pub use aws::S3Storage;
+#[cfg(feature = "cloud-gcp")]
+pub use gcp::GCSStorage;
+
+#[cfg(feature = "prost-codec")]
+pub use kvproto::backup::storage_backend::Backend;
+#[cfg(feature = "protobuf-codec")]
+pub use kvproto::backup::StorageBackend_oneof_backend as Backend;
+use kvproto::backup::{CloudDynamic, Gcs, S3};
+
+use cloud::blob::BlobStorage;
+pub use external_storage::{record_storage_create, ExternalStorage, LocalStorage, NoopStorage};
+use futures_io::AsyncRead;
+use kvproto::backup::{Noop, StorageBackend};
+
+pub fn create_storage(storage_backend: &StorageBackend) -> io::Result<Box<dyn ExternalStorage>> {
+    if let Some(backend) = &storage_backend.backend {
+        create_backend(backend)
+    } else {
+        Err(io::Error::new(
+            io::ErrorKind::NotFound,
+            format!("no storage backend {:?}", storage_backend),
+        ))
+    }
+}
+
+#[cfg(any(feature = "cloud-gcp", feature = "cloud-aws"))]
+fn blob_store(store: Box<dyn BlobStorage>) -> Box<dyn ExternalStorage> {
```

Comment:
Any difference between BlobStorage and ExternalStorage?

Reply:
Currently there is no difference. This gives us some flexibility to change things in the future. But the real reason for duplicate traits is to avoid crates depending on each other. external_storage is not a dependency of cloud/* and cloud/* is not a dependency of external_storage. This crate, external_storage_export, binds them both together.